diff --git a/bolt/lib/Core/Relocation.cpp b/bolt/lib/Core/Relocation.cpp
index 8788482675a4..2d41428c2cce 100644
--- a/bolt/lib/Core/Relocation.cpp
+++ b/bolt/lib/Core/Relocation.cpp
@@ -100,6 +100,146 @@ static bool isSupportedAArch64(uint32_t Type) {
   }
 }
 
+
+static bool isSupportedARM(uint32_t Type) {
+  switch (Type) {
+  default:
+    return false;
+  case ELF::R_ARM_NONE:
+  case ELF::R_ARM_CALL:
+  case ELF::R_ARM_JUMP24:
+  case ELF::R_ARM_THM_CALL:
+  case ELF::R_ARM_THM_JUMP24:
+  case ELF::R_ARM_ABS32:
+  case ELF::R_ARM_REL32:
+  case ELF::R_ARM_PC24:
+  case ELF::R_ARM_PLT32:
+  case ELF::R_ARM_PREL31:
+  case ELF::R_ARM_MOVW_ABS_NC:
+  case ELF::R_ARM_MOVT_ABS:
+  case ELF::R_ARM_V4BX:
+  case ELF::R_ARM_TARGET1:
+  case ELF::R_ARM_TARGET2:
+    return true;
+  }
+}
+
+static size_t getSizeForTypeARM(uint32_t Type) {
+  switch (Type) {
+  default:
+    errs() << object::getELFRelocationTypeName(ELF::EM_ARM, Type) << '\n';
+    llvm_unreachable("unsupported relocation type");
+  case ELF::R_ARM_ABS32:
+  case ELF::R_ARM_REL32:
+  case ELF::R_ARM_PC24:
+  case ELF::R_ARM_CALL:
+  case ELF::R_ARM_JUMP24:
+  case ELF::R_ARM_PLT32:
+  case ELF::R_ARM_PREL31:
+  case ELF::R_ARM_MOVW_ABS_NC:
+  case ELF::R_ARM_MOVT_ABS:
+    return 4;
+  case ELF::R_ARM_THM_CALL:
+  case ELF::R_ARM_THM_JUMP24:
+    return 4;
+  }
+}
+
+static bool skipRelocationTypeARM(uint32_t Type) {
+  switch (Type) {
+  default:
+    return false;
+  case ELF::R_ARM_NONE:
+  case ELF::R_ARM_V4BX:
+    return true;
+  }
+}
+
+static bool isPCRelativeARM(uint32_t Type) {
+  switch (Type) {
+  default:
+    return false;
+  case ELF::R_ARM_CALL:
+  case ELF::R_ARM_JUMP24:
+  case ELF::R_ARM_THM_CALL:
+  case ELF::R_ARM_THM_JUMP24:
+  case ELF::R_ARM_PC24:
+  case ELF::R_ARM_PLT32:
+  case ELF::R_ARM_PREL31:
+  case ELF::R_ARM_REL32:
+    return true;
+  }
+}
+
+static uint64_t encodeValueARM(uint32_t Type, uint64_t Value, uint64_t PC) {
+  switch (Type) {
+  default:
+    llvm_unreachable("unsupported relocation");
+  case ELF::R_ARM_ABS32:
+  case ELF::R_ARM_MOVW_ABS_NC:
+  case ELF::R_ARM_MOVT_ABS:
+    break;
+  case ELF::R_ARM_REL32:
+  case ELF::R_ARM_PC24:
+  case ELF::R_ARM_CALL:
+  case ELF::R_ARM_JUMP24:
+  case ELF::R_ARM_PLT32:
+  case ELF::R_ARM_PREL31:
+  case ELF::R_ARM_THM_CALL:
+  case ELF::R_ARM_THM_JUMP24:
+    Value -= PC;
+    break;
+  }
+  return Value;
+}
+
+static bool canEncodeValueARM(uint32_t Type, uint64_t Value, uint64_t PC) {
+  switch (Type) {
+  default:
+    return true;
+  case ELF::R_ARM_CALL:
+  case ELF::R_ARM_JUMP24:
+    return isInt<26>(Value - PC);
+  case ELF::R_ARM_THM_CALL:
+  case ELF::R_ARM_THM_JUMP24:
+    return isInt<25>(Value - PC);
+  }
+}
+
+static uint64_t extractValueARM(uint32_t Type, uint64_t Contents,
+                                uint64_t PC) {
+  switch (Type) {
+  default:
+    return Contents;
+  case ELF::R_ARM_ABS32:
+    return Contents;
+  case ELF::R_ARM_REL32:
+  case ELF::R_ARM_PC24:
+  case ELF::R_ARM_PLT32:
+  case ELF::R_ARM_PREL31:
+    return static_cast<int64_t>(PC) + SignExtend64<32>(Contents);
+  case ELF::R_ARM_CALL:
+  case ELF::R_ARM_JUMP24: {
+    uint64_t Imm = Contents & 0x00ffffff;
+    return static_cast<int64_t>(PC) + SignExtend64<26>(Imm << 2);
+  }
+  case ELF::R_ARM_THM_CALL:
+  case ELF::R_ARM_THM_JUMP24: {
+    uint32_t S = (Contents >> 26) & 1;
+    uint32_t J1 = (Contents >> 13) & 1;
+    uint32_t J2 = (Contents >> 11) & 1;
+    uint32_t I1 = J1 ^ S ? 0 : 1;
+    uint32_t I2 = J2 ^ S ? 0 : 1;
+    uint32_t Imm10 = (Contents >> 16) & 0x3ff;
+    uint32_t Imm11 = Contents & 0x7ff;
+    uint64_t Offset = (S << 24) | (I1 << 23) | (I2 << 22) | (Imm10 << 12) |
+                      (Imm11 << 1);
+    return static_cast<int64_t>(PC) + SignExtend64<25>(Offset);
+  }
+  }
+}
+
+
 static bool isSupportedRISCV(uint32_t Type) {
   switch (Type) {
   default:
@@ -731,6 +871,8 @@ bool Relocation::isSupported(uint32_t Type) {
   case Triple::riscv64:
   case Triple::riscv32:
     return isSupportedRISCV(Type);
+  case Triple::arm:
+    return isSupportedARM(Type);
   case Triple::x86_64:
     return isSupportedX86(Type);
   }
@@ -743,6 +885,8 @@ size_t Relocation::getSizeForType(uint32_t Type) {
   case Triple::aarch64:
     return getSizeForTypeAArch64(Type);
   case Triple::riscv64:
+  case Triple::arm:
+    return getSizeForTypeARM(Type);
   case Triple::riscv32:
     return getSizeForTypeRISCV(Type);
   case Triple::x86_64:
@@ -757,6 +901,8 @@ bool Relocation::skipRelocationType(uint32_t Type) {
   case Triple::aarch64:
     return skipRelocationTypeAArch64(Type);
   case Triple::riscv64:
+  case Triple::arm:
+    return skipRelocationTypeARM(Type);
   case Triple::riscv32:
     return skipRelocationTypeRISCV(Type);
   case Triple::x86_64:
@@ -771,6 +917,8 @@ uint64_t Relocation::encodeValue(uint32_t Type, uint64_t Value, uint64_t PC) {
   case Triple::aarch64:
     return encodeValueAArch64(Type, Value, PC);
   case Triple::riscv64:
+  case Triple::arm:
+    return encodeValueARM(Type, Value, PC);
   case Triple::riscv32:
     return encodeValueRISCV(Type, Value, PC);
   case Triple::x86_64:
@@ -785,6 +933,8 @@ bool Relocation::canEncodeValue(uint32_t Type, uint64_t Value, uint64_t PC) {
   case Triple::aarch64:
     return canEncodeValueAArch64(Type, Value, PC);
   case Triple::riscv64:
+  case Triple::arm:
+    return canEncodeValueARM(Type, Value, PC);
   case Triple::riscv32:
     return canEncodeValueRISCV(Type, Value, PC);
   case Triple::x86_64:
@@ -800,6 +950,8 @@ uint64_t Relocation::extractValue(uint32_t Type, uint64_t Contents,
   case Triple::aarch64:
     return extractValueAArch64(Type, Contents, PC);
   case Triple::riscv64:
+  case Triple::arm:
+    return extractValueARM(Type, Contents, PC);
   case Triple::riscv32:
     return extractValueRISCV(Type, Contents, PC);
   case Triple::x86_64:
@@ -897,6 +1049,8 @@ uint32_t Relocation::getNone() {
   case Triple::aarch64:
     return ELF::R_AARCH64_NONE;
   case Triple::riscv64:
+  case Triple::arm:
+    return ELF::R_ARM_NONE;
   case Triple::riscv32:
     return ELF::R_RISCV_NONE;
   case Triple::x86_64:
@@ -911,6 +1065,8 @@ uint32_t Relocation::getPC32() {
   case Triple::aarch64:
     return ELF::R_AARCH64_PREL32;
   case Triple::riscv64:
+  case Triple::arm:
+    return ELF::R_ARM_REL32;
   case Triple::riscv32:
     return ELF::R_RISCV_32_PCREL;
   case Triple::x86_64:
@@ -925,6 +1081,8 @@ uint32_t Relocation::getPC64() {
   case Triple::aarch64:
     return ELF::R_AARCH64_PREL64;
   case Triple::riscv64:
+  case Triple::arm:
+    return ELF::R_ARM_ABS32;
   case Triple::riscv32:
     llvm_unreachable("not implemented");
   case Triple::x86_64:
@@ -945,6 +1103,8 @@ bool Relocation::isPCRelative(uint32_t Type) {
   case Triple::aarch64:
     return isPCRelativeAArch64(Type);
   case Triple::riscv64:
+  case Triple::arm:
+    return isPCRelativeARM(Type);
   case Triple::riscv32:
     return isPCRelativeRISCV(Type);
   case Triple::x86_64:
